File: lc_soln/0084_largest_rectangle_in_histogram/solution.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        std::vector<int> lefts(heights.size(), -1);
        std::stack<int> stack;
        for (int i = 0; i < heights.size(); ++i) {
            while (!stack.empty() && heights[stack.top()] >= heights[i]) {
                stack.pop();
            }
            if (!stack.empty()) {
                lefts[i] = stack.top();
            }
            stack.push(i);
        }
        std::vector<int> rights(heights.size(), heights.size());
        std::stack<int>().swap(stack);
        for (int i = heights.size() - 1; i >= 0; --i) {
            while (!stack.empty() && heights[stack.top()] >= heights[i]) {
                stack.pop();
            }
            if (!stack.empty()) {
                rights[i] = stack.top();
            }
            stack.push(i);
        }
        int result = 0;
        for (int i = 0; i < heights.size(); ++i) {
            result = std::max(result, heights[i] * (rights[i] - lefts[i] - 1));
        }
        return result;
    }
};

int main() {
    return 0;
}

File: lc_soln/0084_largest_rectangle_in_histogram/solution.cs
public class Solution
{
    public int LargestRectangleArea(int[] heights)
    {
        int[] lefts = new int[heights.Length];
        Array.Fill(lefts, -1);
        var stack = new Stack<int>();
        for (int i = 0; i < heights.Length; ++i)
        {
            while (stack.Count > 0 && heights[stack.Peek()] >= heights[i])
            {
                stack.Pop();
            }
            if (stack.Count > 0)
            {
                lefts[i] = stack.Peek();
            }
            stack.Push(i);
        }

        int[] rights = new int[heights.Length];
        Array.Fill(rights, heights.Length);
        stack.Clear();
        for (int i = heights.Length - 1; i >= 0; --i)
        {
            while (stack.Count > 0 && heights[stack.Peek()] >= heights[i])
            {
                stack.Pop();
            }
            if (stack.Count > 0)
            {
                rights[i] = stack.Peek();
            }
            stack.Push(i);
        }

        int result = 0;
        for (int i = 0; i < heights.Length; ++i)
        {
            result = int.Max(result, heights[i] * (rights[i] - lefts[i] - 1));
        }
        return result;
    }
}
